File: Sources/DoNotTypeApp/RecordingOverlay.swift
import AppKit
import DoNotTypeCore
import SwiftUI

/// The floating pill at the bottom of the screen while recording.
///
/// A borderless, non-activating panel: it must never steal focus, because the whole point is that
/// the user keeps typing into whatever app they were already in. `.statusBar` level puts it above
/// full-screen apps, and `canJoinAllSpaces` keeps it visible when they switch desktops mid-thought.
@MainActor
final class RecordingOverlay {
    private var panel: NSPanel?
    private var dismissWork: Task<Void, Never>?
    private let state = OverlayState()

    /// Adds however many bars the microphone has produced since the last redraw.
    func append(levels: [AudioLevelMeter.Bar]) {
        state.append(levels)
    }

    func update(phase: OverlayState.Phase) {
        state.phase = phase
    }

    func show(phase: OverlayState.Phase, hint: String) {
        dismissWork?.cancel()
        state.phase = phase
        state.hint = hint
        state.clearLevels()

        if panel == nil { panel = makePanel() }
        position(panel)
        panel?.orderFrontRegardless()
        // Animated inside SwiftUI rather than on the window: animating an NSPanel's alpha while
        // it is being ordered in produces a visible flash on the first frame.
        withAnimation(.spring(duration: 0.28, bounce: 0.22)) { state.isPresented = true }
    }

    /// Flashes a confirmation, then dismisses itself.
    ///
    /// - Parameter rewriteFailed: the words landed but the rewrite that was asked for did not
    ///   happen. Worth saying: somebody who held the rewrite key and got their own words back
    ///   would otherwise have to notice, and the usual cause is a backend that cannot rewrite at
    ///   all rather than anything they did.
    func confirmInserted(characters: Int, rewriteFailed: Bool = false) {
        update(phase: .inserted(characters, rewriteFailed: rewriteFailed))
        hide(after: rewriteFailed ? .seconds(3) : .milliseconds(900))
    }

    func hide(after delay: Duration = .zero) {
        dismissWork?.cancel()
        let work = Task { @MainActor in
            if delay > .zero { try? await Task.sleep(for: delay) }
            guard !Task.isCancelled else { return }

            withAnimation(.easeIn(duration: 0.18)) { state.isPresented = false }
            // Ordering the panel out before the animation finishes would cut it off mid-fade.
            try? await Task.sleep(for: .milliseconds(200))
            guard !Task.isCancelled else { return }
            panel?.orderOut(nil)
        }
        dismissWork = work
    }

    // MARK: - Private

    private func makePanel() -> NSPanel {
        // Wider than any pill it will hold, rather than the size of one.
        //
        // A window clips its content to its own frame, and the pill sizes itself to whatever it is
        // currently saying — which at 220 points was routinely more than it had room for. SwiftUI
        // does not overflow: it wraps, and then it truncates. So the effect was invisible in the
        // code and plain on screen. "Release or tap to send" broke onto two lines, and a failure
        // message reached the two-line limit and ended in an ellipsis, which is the one thing this
        // project does not do to an error — that text is what somebody copies to find out what
        // went wrong.
        //
        // Transparent and mouse-ignoring, so the room it does not draw in costs nothing, and the
        // pill goes on centring itself inside it.
        let panel = NSPanel(
            contentRect: NSRect(x: 0, y: 0, width: 460, height: 64),
            styleMask: [.borderless, .nonactivatingPanel],
            backing: .buffered, defer: false)

        panel.isFloatingPanel = true
        panel.level = .statusBar
        panel.backgroundColor = .clear
        panel.isOpaque = false
        panel.hasShadow = true
        panel.ignoresMouseEvents = true
        panel.collectionBehavior = [.canJoinAllSpaces, .fullScreenAuxiliary, .stationary]
        // Without this the panel takes key focus and the user's caret goes away.
        panel.hidesOnDeactivate = false

        let hosting = NSHostingView(rootView: OverlayView(state: state))
        hosting.frame = panel.contentRect(forFrameRect: panel.frame)
        panel.contentView = hosting
        return panel
    }

    /// Bottom-centre of whichever screen has the mouse, so it follows a multi-display setup.
    private func position(_ panel: NSPanel?) {
        guard let panel else { return }
        let mouse = NSEvent.mouseLocation
        let screen = NSScreen.screens.first { NSMouseInRect(mouse, $0.frame, false) }
            ?? NSScreen.main
        guard let frame = screen?.visibleFrame else { return }

        let size = panel.frame.size
        panel.setFrameOrigin(
            NSPoint(
                x: frame.midX - size.width / 2,
                y: frame.minY + 64))
    }
}

@MainActor
@Observable
final class OverlayState {
    enum Phase: Equatable {
        case recording
        case transcribing
        /// A long dictation split across several requests. Shown only when there is more than one,
        /// because "1 of 1" is noise — but a nine-minute recording that sits on "Transcribing…"
        /// for half a minute looks hung, and this is what distinguishes slow from stuck.
        case transcribingChunk(done: Int, of: Int)
        /// The second request, which is a different thing being waited on and often the slower of
        /// the two. This said "Transcribing…" while a model rewrote a transcript that was already
        /// finished — so the one phase where the wait is the *model's* thinking was the one phase
        /// that did not say so.
        case deriving(RewriteStyle)
        /// Brief confirmation that words were inserted, so success is visible rather than a
        /// silent disappearance the user has to infer from the text appearing.
        case inserted(Int, rewriteFailed: Bool)
        case failed(String)
    }

    /// How much of the recording the meter shows: 24 bars of 60 ms, so a second and a half.
    ///
    /// Long enough that the sentence you are in the middle of saying is on screen, short enough
    /// that the bars stay wide enough to read.
    static let visibleBars = 24

    var phase: Phase = .recording
    var hint: String = ""
    /// The visible history, oldest first. Always full: the meter starts flat rather than growing
    /// in from the left, because an empty meter and a silent one should not look different.
    private(set) var levels = [AudioLevelMeter.Bar](repeating: .silent, count: visibleBars)
    /// Drives the appear/disappear transition.
    var isPresented = false

    func append(_ bars: [AudioLevelMeter.Bar]) {
        guard !bars.isEmpty else { return }
        levels = Array((levels + bars).suffix(Self.visibleBars))
    }

    func clearLevels() {
        levels = [AudioLevelMeter.Bar](repeating: .silent, count: Self.visibleBars)
    }
}

private struct OverlayView: View {
    @Bindable var state: OverlayState
    @Environment(\.accessibilityReduceMotion) private var reduceMotion


    var body: some View {
        content
            // A pill that scales and fades in reads as "this appeared for you"; one that pops
            // reads as a glitch. Respecting reduce-motion is not optional for something that
            // shows up unannounced in the corner of the screen.
            .opacity(state.isPresented ? 1 : 0)
            .scaleEffect(reduceMotion ? 1 : (state.isPresented ? 1 : 0.88))
            .offset(y: reduceMotion ? 0 : (state.isPresented ? 0 : 12))
    }

    private var content: some View {
        HStack(spacing: 12) {
            switch state.phase {
            case .recording:
                LevelMeter(bars: state.levels)
                    .frame(width: LevelMeter.width, height: 22)
                Text(state.hint)
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(.white.opacity(0.75))
            case .transcribing:
                ThinkingDots()
                Text("Transcribing…")
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(.white.opacity(0.75))
            case .transcribingChunk(let done, let total):
                ProgressView(value: Double(done), total: Double(max(total, 1)))
                    .progressViewStyle(.linear)
                    .tint(.white)
                    .frame(width: LevelMeter.width)
                    // A determinate bar here, not the dots: with several parts in flight there is
                    // real progress to report, and reporting it beats implying it.
                Text("Transcribing… part \(min(done + 1, total)) of \(total)")
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(.white.opacity(0.75))
                    .monospacedDigit()
            case .deriving(let style):
                ThinkingDots()
                Text(TranscriptMode.rewrite(style).progressLabel)
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(.white.opacity(0.75))
            case .inserted(let characters, let rewriteFailed):
                Image(systemName: rewriteFailed
                    ? "exclamationmark.circle.fill" : "checkmark.circle.fill")
                    .foregroundStyle(rewriteFailed ? .orange : .green)
                    .transition(.scale.combined(with: .opacity))
                Text(
                    "Inserted \(characters) character\(characters == 1 ? "" : "s")"
                        + (rewriteFailed ? " — not rewritten" : ""))
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(.white.opacity(0.75))
            case .failed(let message):
                Image(systemName: "exclamationmark.triangle.fill")
                    .foregroundStyle(.orange)
                Text(message)
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(.white.opacity(0.85))
                    .lineLimit(2)
                    .frame(maxWidth: 320, alignment: .leading)
            }
        }
        .padding(.horizontal, 18)
        .frame(minHeight: 44)
        .animation(.easeInOut(duration: 0.2), value: state.phase)
        .background(
            Capsule().fill(Color.black.opacity(0.82))
        )
        .overlay(
            Capsule().strokeBorder(.white.opacity(0.12), lineWidth: 1)
        )
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

/// The last second and a half of the microphone, walking leftwards.
///
/// What replaced it is worth saying, because it looked like this one. The old meter was five bars
/// driven by a single current level, animated by a travelling sine so that they kept moving during
/// pauses. Two things were wrong with that. The level it drew was `min(1, rms * 6)`, which spent
/// most of a normally-recorded voice pinned at full height — see `AudioLevelMeter` for the
/// measurement — so it could report that sound was arriving but never how much. And the movement
/// was invented: the bars swayed identically whether the mic was hearing a sentence or nothing at
/// all, which is the one question somebody looks at a level meter to answer.
///
/// Here every bar is 60 ms of audio that actually happened, and the meter moves because the audio
/// does. Silence is a flat line still scrolling: the mic is live and hearing nothing. That reads as
/// a report rather than as decoration, and it is why this stays animated under Reduce Motion when
/// the transcribing dots do not — the motion is the measurement.
private struct LevelMeter: View {
    let bars: [AudioLevelMeter.Bar]

    private static let barWidth = 3.0
    private static let spacing = 2.0
    /// Sized from the bar count so the pill's width does not depend on two numbers agreeing.
    static var width: Double {
        Double(OverlayState.visibleBars) * barWidth
            + Double(OverlayState.visibleBars - 1) * spacing
    }

    var body: some View {
        GeometryReader { geometry in
            HStack(alignment: .center, spacing: Self.spacing) {
                ForEach(Array(bars.enumerated()), id: \.offset) { _, bar in
                    Capsule()
                        // Amber is not decoration: the input is loud enough to be clipped on the
                        // way in, and a recording distorted before it is sent is worth one colour.
                        .fill(bar.isClipping ? Color.orange : .white.opacity(0.9))
                        .frame(width: Self.barWidth, height: height(bar, in: geometry.size.height))
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            // One bar's worth, so the uneven arrival of buffers does not show up as a stutter.
            // Longer than this and the waveform smears into a blur that flatters the signal.
            .animation(.linear(duration: 0.05), value: bars)
        }
    }

    /// Silence is a row of dots rather than nothing at all: a meter that disappears when the room
    /// is quiet cannot be told apart from one that has stopped.
    private func height(_ bar: AudioLevelMeter.Bar, in available: Double) -> Double {
        max(Self.barWidth, available * bar.level)
    }
}

/// The "still working" animation, shown once speech has stopped.
///
/// Deliberately unlike the recording waveform, which is driven by the microphone. After the user
/// stops talking there is no input left to reflect, so a level-driven animation would be
/// decoration pretending to be a signal. What this has to convey is only "not hung" — the failure
/// it prevents is someone deciding nothing happened and pressing the key again mid-request.
///
/// Occupies the same width as the waveform so the pill does not resize between phases; a capsule
/// that jumps as it changes state reads as a glitch.
private struct ThinkingDots: View {
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        TimelineView(.animation(minimumInterval: 1.0 / 30.0, paused: reduceMotion)) { timeline in
            let time = timeline.date.timeIntervalSinceReferenceDate
            HStack(spacing: 6) {
                ForEach(0..<3, id: \.self) { index in
                    // Each dot lags the one before it, so the group reads as motion in one
                    // direction rather than three things blinking independently.
                    let local = reduceMotion ? 0.5 : abs(sin(time * 3 - Double(index) * 0.7))
                    Circle()
                        .fill(.white.opacity(0.45 + 0.5 * local))
                        .frame(width: 6 + 3 * local, height: 6 + 3 * local)
                }
            }
            .frame(width: LevelMeter.width, alignment: .center)
        }
    }
}

File: windows/DoNotType.App/RecordingOverlay.cs
using System.Drawing.Drawing2D;
using System.Runtime.InteropServices;
using System.Windows.Forms;
using DoNotType.Core;

namespace DoNotType.App;

/// <summary>
/// The floating pill at the bottom of the screen while recording.
///
/// It must never take focus, because the whole point is that the caret stays in whatever app the
/// user was already typing in. <c>WS_EX_NOACTIVATE</c> plus <see cref="ShowWithoutActivation"/>
/// is what guarantees that; a plain topmost form would steal it.
/// </summary>
public sealed class RecordingOverlay : Form
{
    public enum Phase
    {
        Recording,
        Transcribing,
        /// <summary>
        /// The second request. Distinct from <see cref="Transcribing"/> because it is a different
        /// thing to be waiting on and usually the slower of the two — and because the transcript
        /// exists by then, so nothing on screen is moving and a stale label reads as a hang.
        /// </summary>
        Deriving,
        /// <summary>
        /// Brief confirmation that words were inserted, so success is visible rather than a silent
        /// disappearance the user has to infer from the text appearing.
        /// </summary>
        Inserted,
        Failed,
    }

    private const int WS_EX_NOACTIVATE = 0x08000000;
    private const int WS_EX_TOOLWINDOW = 0x00000080;
    private const int WS_EX_TRANSPARENT = 0x00000020;

    /// <summary>
    /// How much of the recording the meter shows: 24 bars of 60 ms, so a second and a half. Long
    /// enough that the sentence being spoken is on screen, short enough for the bars to stay
    /// readable.
    /// </summary>
    private const int VisibleBars = 24;
    private const int MeterBarWidth = 3;
    private const int MeterBarGap = 2;
    /// <summary>Sized from the bar count, so the pill's width follows from one number.</summary>
    private const int MeterWidth = (VisibleBars * MeterBarWidth) + ((VisibleBars - 1) * MeterBarGap);

    /// <summary>
    /// The visible history, oldest first. Always full: the meter starts flat rather than growing in
    /// from the left, because an empty meter and a silent one should not look different.
    /// </summary>
    private readonly AudioLevelMeter.Bar[] _bars =
        [.. Enumerable.Repeat(AudioLevelMeter.Bar.Silent, VisibleBars)];

    private readonly System.Windows.Forms.Timer _animation = new() { Interval = 33 };
    /// <summary>Room for the meter and a few words. Everything but a failure fits.</summary>
    private const int CompactWidth = 320;
    private const int CompactHeight = 52;

    /// <summary>
    /// Wider, and as tall as the text needs, for a failure.
    /// </summary>
    /// <remarks>
    /// The message used to be cut to 48 characters by the caller, which is shorter than every
    /// sentence worth reading — "The API key was rejected. Check it in Settings." does not fit, and
    /// what somebody saw was the first half of the diagnosis with the instruction missing.
    /// </remarks>
    private const int FailureWidth = 420;

    private Phase _phase = Phase.Recording;
    private string _hint = string.Empty;

    public RecordingOverlay()
    {
        FormBorderStyle = FormBorderStyle.None;
        ShowInTaskbar = false;
        TopMost = true;
        StartPosition = FormStartPosition.Manual;
        Size = new Size(CompactWidth, CompactHeight);
        BackColor = Color.Black;
        // Rounded pill: the region is cheaper and crisper than an alpha-blended layered window.
        DoubleBuffered = true;

        // Repaints the thinking dots, which are drawn from the clock. The meter does not need it:
        // it repaints when audio arrives, which is the only time it has anything new to say.
        _animation.Tick += (_, _) => Invalidate();
    }

    protected override bool ShowWithoutActivation => true;

    protected override CreateParams CreateParams
    {
        get
        {
            var parameters = base.CreateParams;
            parameters.ExStyle |= WS_EX_NOACTIVATE | WS_EX_TOOLWINDOW | WS_EX_TRANSPARENT;
            return parameters;
        }
    }

    /// <summary>Adds however many bars the microphone has produced since the last redraw.</summary>
    public void AppendLevels(IReadOnlyList<AudioLevelMeter.Bar> bars)
    {
        if (bars.Count == 0) return;

        var incoming = Math.Min(bars.Count, VisibleBars);
        Array.Copy(_bars, incoming, _bars, 0, VisibleBars - incoming);
        for (var index = 0; index < incoming; index++)
        {
            _bars[VisibleBars - incoming + index] = bars[bars.Count - incoming + index];
        }
        Invalidate();
    }

    private void ClearLevels() => Array.Fill(_bars, AudioLevelMeter.Bar.Silent);

    public void Show(Phase phase, string hint)
    {
        _phase = phase;
        _hint = hint;
        ClearLevels();
        ResizeForPhase();
        PositionAtBottomOfActiveScreen();
        if (!Visible) base.Show();
        _animation.Start();
        Invalidate();
    }

    /// <param name="hint">Extra detail under the phase, or null when there is none worth showing.</param>
    public void SetPhase(Phase phase, string? hint)
    {
        _phase = phase;
        _hint = hint ?? string.Empty;
        ResizeForPhase();
        PositionAtBottomOfActiveScreen();
        Invalidate();
    }

    /// <summary>
    /// A failure gets as much room as its sentence needs; everything else stays a pill.
    /// </summary>
    private void ResizeForPhase()
    {
        if (_phase != Phase.Failed)
        {
            Size = new Size(CompactWidth, CompactHeight);
            return;
        }

        using var graphics = CreateGraphics();
        using var font = MessageFont();
        var measured = graphics.MeasureString(
            _hint, font, FailureWidth - TextLeft - 24);
        Size = new Size(FailureWidth, Math.Max(CompactHeight, (int)measured.Height + 28));
    }

    public new void Hide()
    {
        _animation.Stop();
        base.Hide();
    }

    /// <summary>Bottom-centre of whichever screen has the cursor, so it follows a multi-monitor setup.</summary>
    private void PositionAtBottomOfActiveScreen()
    {
        var screen = Screen.FromPoint(Cursor.Position).WorkingArea;
        Location = new Point(
            screen.Left + (screen.Width - Width) / 2,
            screen.Bottom - Height - 64);
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        var g = e.Graphics;
        g.SmoothingMode = SmoothingMode.AntiAlias;
        g.Clear(Color.Black);

        using var path = RoundedRect(new Rectangle(0, 0, Width - 1, Height - 1), Height / 2);
        using var fill = new SolidBrush(Color.FromArgb(220, 18, 18, 20));
        using var border = new Pen(Color.FromArgb(40, 255, 255, 255));
        g.FillPath(fill, path);
        g.DrawPath(border, path);
        Region = new Region(path);

        using var textBrush = new SolidBrush(Color.FromArgb(200, 235, 235, 235));
        using var font = MessageFont();

        switch (_phase)
        {
            case Phase.Recording:
                DrawMeter(g, new Rectangle(20, Height / 2 - 11, MeterWidth, 22));
                g.DrawString(_hint, font, textBrush, TextLeft, Height / 2 - 8);
                break;

            case Phase.Transcribing:
                // Dots rather than a static label alone. After the user stops talking the wait is
                // dead time, and the failure to prevent is them deciding nothing happened and
                // pressing the key again mid-request. Deliberately unlike the recording waveform:
                // there is no input left to reflect, so anything level-driven would be decoration
                // pretending to be a signal.
                DrawThinkingDots(g, new Rectangle(20, Height / 2 - 4, MeterWidth, 8));
                var label = _hint.Length == 0 ? "Transcribing…" : $"Transcribing… {_hint}";
                g.DrawString(label, font, textBrush, TextLeft, Height / 2 - 8);
                break;

            case Phase.Deriving:
                DrawThinkingDots(g, new Rectangle(20, Height / 2 - 4, MeterWidth, 8));
                g.DrawString(
                    _hint.Length == 0 ? "Rewriting…" : _hint, font, textBrush,
                    TextLeft, Height / 2 - 8);
                break;

            case Phase.Inserted:
                DrawTick(g, new Rectangle(24, Height / 2 - 7, 14, 14));
                g.DrawString(_hint, font, textBrush, 48, Height / 2 - 8);
                break;

            default:
                using (var warn = new SolidBrush(Color.FromArgb(230, 240, 160, 90)))
                using (var format = new StringFormat { Trimming = StringTrimming.EllipsisWord })
                {
                    // Wrapped rather than cut. The advice is a sentence about what to do, and half
                    // a sentence about what to do is worse than none.
                    g.DrawString(
                        _hint, font, warn,
                        new RectangleF(24, 14, Width - 48, Height - 24), format);
                }
                break;
        }
    }

    private static Font MessageFont() =>
        new(SystemFonts.MessageBoxFont!.FontFamily, 9f, FontStyle.Regular);

    /// <summary>Where the label starts, clear of the meter or the dots.</summary>
    private const int TextLeft = 20 + MeterWidth + 12;

    /// <summary>A tick, drawn rather than shipped as an icon so it inherits the pill's colours.</summary>
    private static void DrawTick(Graphics g, Rectangle bounds)
    {
        using var pen = new Pen(Color.FromArgb(230, 120, 210, 130), 2.2f)
        {
            StartCap = LineCap.Round,
            EndCap = LineCap.Round,
        };
        g.DrawLines(pen, [
            new PointF(bounds.Left, bounds.Top + bounds.Height * 0.55f),
            new PointF(bounds.Left + bounds.Width * 0.38f, bounds.Bottom - 1),
            new PointF(bounds.Right, bounds.Top + 1),
        ]);
    }

    /// <summary>Three dots travelling in one direction, so the group reads as motion rather than
    /// as three things blinking independently.</summary>
    private void DrawThinkingDots(Graphics g, Rectangle bounds)
    {
        const int dots = 3;
        var phase = Environment.TickCount / 1000.0;
        var spacing = bounds.Width / (float)(dots + 1);

        for (var index = 0; index < dots; index++)
        {
            var local = Math.Abs(Math.Sin(phase * 3 - index * 0.7));
            var radius = (float)(3 + 2 * local);
            var alpha = (int)Math.Clamp(120 + 135 * local, 0, 255);

            using var brush = new SolidBrush(Color.FromArgb(alpha, 127, 178, 255));
            var centreX = bounds.Left + spacing * (index + 1);
            g.FillEllipse(
                brush, centreX - radius, bounds.Top + bounds.Height / 2f - radius,
                radius * 2, radius * 2);
        }
    }

    /// <summary>
    /// The last second and a half of the microphone, walking leftwards.
    /// </summary>
    /// <remarks>
    /// <para>
    /// What it replaced is worth saying, because it looked like this. The old meter was five bars
    /// driven by a single current level and animated by a travelling sine wave, and neither half
    /// reported anything: the level was <c>peak * 6</c> clamped, which a normally-recorded voice
    /// sat pinned at the top of — see <see cref="AudioLevelMeter"/> for the measurement — and the
    /// movement was invented, so the bars swayed identically whether the microphone was hearing a
    /// sentence or nothing at all.
    /// </para>
    /// <para>
    /// Here every bar is 60 ms of audio that actually happened, and the meter moves because the
    /// audio does. Silence is a flat row of dots that keeps scrolling: the microphone is live and
    /// hearing nothing.
    /// </para>
    /// </remarks>
    private void DrawMeter(Graphics g, Rectangle bounds)
    {
        using var brush = new SolidBrush(Color.FromArgb(230, 245, 245, 245));
        // Amber is not decoration: the input is loud enough to be clipped on the way in, and a
        // recording distorted before it is sent is worth one colour. The same amber as a failure,
        // because it is the same kind of news.
        using var clipped = new SolidBrush(Color.FromArgb(230, 240, 160, 90));

        for (var index = 0; index < _bars.Length; index++)
        {
            var bar = _bars[index];
            // Silence is a row of dots rather than nothing at all: a meter that disappears when the
            // room is quiet cannot be told apart from one that has stopped.
            var height = Math.Max(MeterBarWidth, (int)(bounds.Height * bar.Level));
            var x = bounds.Left + index * (MeterBarWidth + MeterBarGap);
            var y = bounds.Top + (bounds.Height - height) / 2;

            using var shape = RoundedRect(
                new Rectangle(x, y, MeterBarWidth, height), MeterBarWidth / 2);
            g.FillPath(bar.IsClipping ? clipped : brush, shape);
        }
    }

    private static GraphicsPath RoundedRect(Rectangle bounds, int radius)
    {
        var path = new GraphicsPath();
        var diameter = Math.Max(1, radius * 2);
        path.AddArc(bounds.Left, bounds.Top, diameter, diameter, 180, 90);
        path.AddArc(bounds.Right - diameter, bounds.Top, diameter, diameter, 270, 90);
        path.AddArc(bounds.Right - diameter, bounds.Bottom - diameter, diameter, diameter, 0, 90);
        path.AddArc(bounds.Left, bounds.Bottom - diameter, diameter, diameter, 90, 90);
        path.CloseFigure();
        return path;
    }

    protected override void Dispose(bool disposing)
    {
        if (disposing) _animation.Dispose();
        base.Dispose(disposing);
    }
}
